File: gestalt-core/src/test/java/org/github/gestalt/config/decoder/ByteDecoderTest.java
package org.github.gestalt.config.decoder;

import org.github.gestalt.config.entity.ValidationLevel;
import org.github.gestalt.config.exceptions.GestaltException;
import org.github.gestalt.config.lexer.SentenceLexer;
import org.github.gestalt.config.node.ConfigNodeService;
import org.github.gestalt.config.node.LeafNode;
import org.github.gestalt.config.reflect.TypeCapture;
import org.github.gestalt.config.utils.ValidateOf;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.nio.charset.Charset;
import java.util.Collections;
import java.util.List;

class ByteDecoderTest {

    ConfigNodeService configNodeService;
    SentenceLexer lexer;

    @BeforeEach
    void setup() {
        configNodeService = Mockito.mock(ConfigNodeService.class);
        lexer = Mockito.mock(SentenceLexer.class);
    }

    @Test
    void name() {
        ByteDecoder decoder = new ByteDecoder();
        Assertions.assertEquals("Byte", decoder.name());
    }

    @Test
    void priority() {
        ByteDecoder decoder = new ByteDecoder();
        Assertions.assertEquals(Priority.MEDIUM, decoder.priority());
    }

    @Test
    void matches() {
        ByteDecoder decoder = new ByteDecoder();

        Assertions.assertTrue(decoder.matches(TypeCapture.of(Byte.class)));
        Assertions.assertTrue(decoder.matches(new TypeCapture<Byte>() {
        }));
        Assertions.assertTrue(decoder.matches(TypeCapture.of(byte.class)));

        Assertions.assertFalse(decoder.matches(TypeCapture.of(Character.class)));
        Assertions.assertFalse(decoder.matches(TypeCapture.of(Float.class)));
        Assertions.assertFalse(decoder.matches(TypeCapture.of(String.class)));
        Assertions.assertFalse(decoder.matches(TypeCapture.of(Integer.class)));
        Assertions.assertFalse(decoder.matches(new TypeCapture<List<Float>>() {
        }));
    }

    @Test
    void decodeByte() throws GestaltException {
        ByteDecoder decoder = new ByteDecoder();

        ValidateOf<Byte> validate = decoder.decode("db.port", new LeafNode("a"), TypeCapture.of(Byte.class),
            new DecoderRegistry(Collections.singletonList(decoder), configNodeService, lexer));
        Assertions.assertTrue(validate.hasResults());
        Assertions.assertFalse(validate.hasErrors());
        Assertions.assertEquals("a".getBytes(Charset.defaultCharset())[0], validate.results());
        Assertions.assertEquals(0, validate.getErrors().size());
    }

    @Test
    void notAByteTooLong() throws GestaltException {
        ByteDecoder decoder = new ByteDecoder();

        ValidateOf<Byte> validate = decoder.decode("db.port", new LeafNode("aaa"), TypeCapture.of(Byte.class),
            new DecoderRegistry(Collections.singletonList(decoder), configNodeService, lexer));
        Assertions.assertFalse(validate.hasResults());
        Assertions.assertTrue(validate.hasErrors());
        Assertions.assertNull(validate.results());
        Assertions.assertNotNull(validate.getErrors());
        Assertions.assertEquals(ValidationLevel.WARN, validate.getErrors().get(0).level());
        Assertions.assertEquals("Expected a Byte on path: db.port, decoding node: LeafNode{value='aaa'} received the wrong size",
            validate.getErrors().get(0).description());
    }

    @Test
    void notAByteTooShort() throws GestaltException {
        ByteDecoder decoder = new ByteDecoder();

        ValidateOf<Byte> validate = decoder.decode("db.port", new LeafNode(""), TypeCapture.of(Byte.class),
            new DecoderRegistry(Collections.singletonList(decoder), configNodeService, lexer));
        Assertions.assertFalse(validate.hasResults());
        Assertions.assertTrue(validate.hasErrors());
        Assertions.assertNull(validate.results());
        Assertions.assertNotNull(validate.getErrors());
        Assertions.assertEquals(ValidationLevel.WARN, validate.getErrors().get(0).level());
        Assertions.assertEquals("Expected a Byte on path: db.port, decoding node: LeafNode{value=''} received the wrong size",
            validate.getErrors().get(0).description());
    }
}

File: gestalt-kotlin/src/test/kotlin/org/config/gestalt/kotlin/decoder/ByteDecoderTest.kt
package org.config.gestalt.kotlin.decoder

import org.config.gestalt.decoder.DecoderRegistry
import org.config.gestalt.entity.ValidationLevel
import org.config.gestalt.exceptions.GestaltException
import org.config.gestalt.kotlin.reflect.kTypeCaptureOf
import org.config.gestalt.lexer.SentenceLexer
import org.config.gestalt.node.ConfigNodeService
import org.config.gestalt.node.LeafNode
import org.config.gestalt.reflect.TypeCapture
import org.config.gestalt.utils.ValidateOf
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.nio.charset.Charset
import java.util.*

internal class ByteDecoderTest {
    var configNodeService: ConfigNodeService? = null
    var lexer: SentenceLexer? = null

    @BeforeEach
    fun setup() {
        configNodeService = Mockito.mock(ConfigNodeService::class.java)
        lexer = Mockito.mock(SentenceLexer::class.java)
    }

    @Test
    fun name() {
        val decoder = ByteDecoder()
        Assertions.assertEquals("Byte", decoder.name())
    }

    @Test
    fun matches() {
        val decoder = ByteDecoder()
        Assertions.assertTrue(decoder.matches(kTypeCaptureOf<Byte>()))
        Assertions.assertFalse(decoder.matches(object : TypeCapture<Byte?>() {}))
        Assertions.assertFalse(decoder.matches(TypeCapture.of(Byte::class.java)))
        Assertions.assertFalse(decoder.matches(kTypeCaptureOf<Int>()))
        Assertions.assertFalse(decoder.matches(kTypeCaptureOf<String>()))
        Assertions.assertFalse(decoder.matches(kTypeCaptureOf<Date>()))
        Assertions.assertFalse(decoder.matches(kTypeCaptureOf<List<Byte>>()))
    }

    @Test
    @Throws(GestaltException::class)
    fun decodeByte() {
        val decoder = ByteDecoder()
        val validate: ValidateOf<Byte> = decoder.decode(
            "db.port", LeafNode("a"), TypeCapture.of(
                Byte::class.java
            ),
            DecoderRegistry(listOf(decoder), configNodeService, lexer)
        )
        Assertions.assertTrue(validate.hasResults())
        Assertions.assertFalse(validate.hasErrors())
        Assertions.assertEquals("a".toByteArray(Charset.defaultCharset())[0], validate.results())
        Assertions.assertEquals(0, validate.errors.size)
    }

    @Test
    @Throws(GestaltException::class)
    fun notAByteTooLong() {
        val decoder = ByteDecoder()
        val validate: ValidateOf<Byte> = decoder.decode(
            "db.port", LeafNode("aaa"), TypeCapture.of(
                Byte::class.java
            ),
            DecoderRegistry(listOf(decoder), configNodeService, lexer)
        )
        Assertions.assertFalse(validate.hasResults())
        Assertions.assertTrue(validate.hasErrors())
        Assertions.assertNull(validate.results())
        Assertions.assertNotNull(validate.errors)
        Assertions.assertEquals(ValidationLevel.WARN, validate.errors[0].level())
        Assertions.assertEquals(
            "Expected a Byte on path: db.port, decoding node: LeafNode{value='aaa'} received the wrong size",
            validate.errors[0].description()
        )
    }

    @Test
    @Throws(GestaltException::class)
    fun notAByteTooShort() {
        val decoder = ByteDecoder()
        val validate: ValidateOf<Byte> = decoder.decode(
            "db.port", LeafNode(""), TypeCapture.of(
                Byte::class.java
            ),
            DecoderRegistry(listOf(decoder), configNodeService, lexer)
        )
        Assertions.assertFalse(validate.hasResults())
        Assertions.assertTrue(validate.hasErrors())
        Assertions.assertNull(validate.results())
        Assertions.assertNotNull(validate.errors)
        Assertions.assertEquals(ValidationLevel.WARN, validate.errors[0].level())
        Assertions.assertEquals(
            "Expected a Byte on path: db.port, decoding node: LeafNode{value=''} received the wrong size",
            validate.errors[0].description()
        )
    }
}
